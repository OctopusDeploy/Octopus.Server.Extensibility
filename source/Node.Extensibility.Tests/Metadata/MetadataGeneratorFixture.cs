using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;
using Assent;
using Newtonsoft.Json;
using NUnit.Framework;
using Octopus.Data.Resources;
using Octopus.Data.Resources.Attributes;
using Octopus.Node.Extensibility.Metadata;

namespace Node.Extensibility.Tests.Metadata
{
    [TestFixture]
    public class MetadataGeneratorFixture
    {
        [Test]
        public void SettingsMetadataShouldBeCorrect()
        {
            IGenerateMetadata generator = new MetadataGenerator();

            var metadata = generator.GetMetadata(typeof(TopLevelResource));

            var serializerSettings = new JsonSerializerSettings
            {
                Formatting = Formatting.Indented,
            };

            var json = JsonConvert.SerializeObject(metadata, serializerSettings);

            this.Assent(json);

        }

        [Test]
        public void MetadataShouldHandleSelfReferences()
        {
            IGenerateMetadata generator = new MetadataGenerator();

            var metadata = generator.GetMetadata<SelfReferencingResource>();

            var serializerSettings = new JsonSerializerSettings
            {
                Formatting = Formatting.Indented,
            };

            var json = JsonConvert.SerializeObject(metadata, serializerSettings);

            this.Assent(json);
        }

        [Test]
        public void MetadataShouldHandleNavigationalProperties()
        {
            IGenerateMetadata generator = new MetadataGenerator();

            var metadata = generator.GetMetadata<ParentResource>();

            var serializerSettings = new JsonSerializerSettings
            {
                Formatting = Formatting.Indented,
            };

            var json = JsonConvert.SerializeObject(metadata, serializerSettings);

            this.Assent(json);
        }

        public void SettingsValuesShouldBeCorrect()
        {
            var resource = new TopLevelResource()
            {
                SecondLevelResource = new SecondLevelResource()
                {
                    SensitiveStringProperty = "String value",
                    BoolProperty = false,
                    IntArrayProperty = new[] { 1, 2, 3 },
                    NullableDateTimeOffsetProperty = null,
                    StringArrayProperty = new[] { "first", "second" },
                },
                DateTimeOffsetProperty = DateTime.Now,
                IntProperty = 4,
                NullableDateTimeProperty = null,
                NullableIntProperty = 5,
                ListOfStringProperty = new List<string> { "1st", "2nd", "3rd" },
            };
        }

        public enum TestEnum
        {
            First = 5,
            [System.ComponentModel.Description("2nd")]
            Second = 7,
            Third = 11,
            Fourth
        }

        public abstract class SettingsResource 
        {
            protected SettingsResource()
            {
                Id = GetType().Name;
            }

            public string Id { get; set; }

            public string SomeValue { get; set; }
        }

        public class SelfReferencingResource : SettingsResource
        {
            public string Name { get; set; }

            public SelfReferencingResource SelfReference { get; set; }
        }

        public class ParentResource : SettingsResource
        {
            public List<ChildResource> Children { get; set; }
        }

        public class ChildResource
        {
            public ParentResource Parent { get; set; }

            public int ChildIntProperty { get; set; }
        }


        public class TopLevelResource : SettingsResource
        {
            public SecondLevelResource SecondLevelResource { get; set; }

            [DisplayName("Duplicated 2nd Level")]
            [System.ComponentModel.Description("This 2nd-level resource has been duplicated")]
            public SecondLevelResource DuplicateSecondLevelResource { get; set; }

            [Required]
            public DateTime? NullableDateTimeProperty { get; set; }

            public DateTimeOffset DateTimeOffsetProperty { get; set; }

            public int IntProperty { get; set; }

            [HasOptions(SelectMode.Single)]
            public TestEnum EnumProp { get; set; }

            public int? NullableIntProperty { get; set; }

            public List<string> ListOfStringProperty { get; set; }

            public DateTime?[] NullableDateTimeArray { get; set; }

            [ReadOnly(true)]
            public string HandsOff { get; set; }
        }

        public class SecondLevelResource
        { 
            public SensitiveValue SensitiveStringProperty { get; set; }

            public DateTimeOffset? NullableDateTimeOffsetProperty { get; set; }

            public bool BoolProperty { get; set; }

            public string[] StringArrayProperty { get; set; }

            public int[] IntArrayProperty { get; set; }
        }
    }
}
