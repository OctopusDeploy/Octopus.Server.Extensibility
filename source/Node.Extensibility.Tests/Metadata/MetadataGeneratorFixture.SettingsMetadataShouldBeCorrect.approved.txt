{
  "Types": [
    {
      "Name": "TopLevelResource",
      "Properties": [
        {
          "Name": "SecondLevelResource",
          "Type": "SecondLevelResource",
          "DisplayInfo": {
            "Required": false,
            "Sensitive": false,
            "Label": "SecondLevelResource",
            "Description": null,
            "Options": null,
            "ListApi": null
          }
        },
        {
          "Name": "DuplicateSecondLevelResource",
          "Type": "SecondLevelResource",
          "DisplayInfo": {
            "Required": false,
            "Sensitive": false,
            "Label": "Duplicated 2nd Level",
            "Description": "This 2nd-level resource has been duplicated",
            "Options": null,
            "ListApi": null
          }
        },
        {
          "Name": "NullableDateTimeProperty",
          "Type": "DateTime?",
          "DisplayInfo": {
            "Required": true,
            "Sensitive": false,
            "Label": "NullableDateTimeProperty",
            "Description": null,
            "Options": null,
            "ListApi": null
          }
        },
        {
          "Name": "DateTimeOffsetProperty",
          "Type": "DateTimeOffset",
          "DisplayInfo": {
            "Required": false,
            "Sensitive": false,
            "Label": "DateTimeOffsetProperty",
            "Description": null,
            "Options": null,
            "ListApi": null
          }
        },
        {
          "Name": "IntProperty",
          "Type": "int",
          "DisplayInfo": {
            "Required": false,
            "Sensitive": false,
            "Label": "IntProperty",
            "Description": null,
            "Options": {
              "MultiSelect": true,
              "Values": null
            },
            "ListApi": null
          }
        },
        {
          "Name": "EnumProp",
          "Type": "TestEnum",
          "DisplayInfo": {
            "Required": false,
            "Sensitive": false,
            "Label": "EnumProp",
            "Description": null,
            "Options": {
              "MultiSelect": false,
              "Values": {
                "5": "First",
                "7": "Second",
                "11": "Third",
                "12": "Fourth"
              }
            },
            "ListApi": null
          }
        },
        {
          "Name": "NullableIntProperty",
          "Type": "int?",
          "DisplayInfo": {
            "Required": false,
            "Sensitive": false,
            "Label": "NullableIntProperty",
            "Description": null,
            "Options": null,
            "ListApi": null
          }
        },
        {
          "Name": "ListOfStringProperty",
          "Type": "string[]",
          "DisplayInfo": {
            "Required": false,
            "Sensitive": false,
            "Label": "ListOfStringProperty",
            "Description": null,
            "Options": null,
            "ListApi": null
          }
        },
        {
          "Name": "NullableDateTimeArray",
          "Type": "DateTime?[]",
          "DisplayInfo": {
            "Required": false,
            "Sensitive": false,
            "Label": "NullableDateTimeArray",
            "Description": null,
            "Options": null,
            "ListApi": null
          }
        },
        {
          "Name": "SomeValue",
          "Type": "string",
          "DisplayInfo": {
            "Required": false,
            "Sensitive": false,
            "Label": "SomeValue",
            "Description": null,
            "Options": null,
            "ListApi": null
          }
        }
      ]
    },
    {
      "Name": "SecondLevelResource",
      "Properties": [
        {
          "Name": "StringProperty",
          "Type": "string",
          "DisplayInfo": {
            "Required": false,
            "Sensitive": true,
            "Label": "StringProperty",
            "Description": null,
            "Options": null,
            "ListApi": null
          }
        },
        {
          "Name": "NullableDateTimeOffsetProperty",
          "Type": "DateTimeOffset?",
          "DisplayInfo": {
            "Required": false,
            "Sensitive": false,
            "Label": "NullableDateTimeOffsetProperty",
            "Description": null,
            "Options": null,
            "ListApi": null
          }
        },
        {
          "Name": "BoolProperty",
          "Type": "bool",
          "DisplayInfo": {
            "Required": false,
            "Sensitive": false,
            "Label": "BoolProperty",
            "Description": null,
            "Options": null,
            "ListApi": null
          }
        },
        {
          "Name": "StringArrayProperty",
          "Type": "string[]",
          "DisplayInfo": {
            "Required": false,
            "Sensitive": false,
            "Label": "StringArrayProperty",
            "Description": null,
            "Options": null,
            "ListApi": {
              "MultiSelect": true
            }
          }
        },
        {
          "Name": "IntArrayProperty",
          "Type": "int[]",
          "DisplayInfo": {
            "Required": false,
            "Sensitive": false,
            "Label": "IntArrayProperty",
            "Description": null,
            "Options": null,
            "ListApi": {
              "MultiSelect": false
            }
          }
        }
      ]
    }
  ]
}